rfcs/008-map-indexes.md

```diff
@@ -0,0 +1,202 @@
+# Map indexes RFC
+
+---
+
+name: Formal RFC
+about: Submit a formal Request For Comments for consideration by the team.
+title: ‘Map indexes on FoundationDB’
+labels: rfc, discussion
+assignees: ''
+
+---
+
+## Introduction
+
+This document describes the data model and index management for building and querying map indexes.
+
+## Abstract
+
+Map indexes will have their data model stored in FoundationDB. Each index is grouped via its design doc's view signature. An index will have the index key/value pairs stored, along with the last sequence number from the changes feed used to update the index.
+
+Indexes will be build using the background jobs api, `couch_jobs`, and will use the changes feed. There will be new size limitations on keys (10KB) and values (100KB) that are emitted from a map function.
+
+## Requirements Language
+
+[note]: # " Do not alter the section below. Follow its instructions. "
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this
+document are to be interpreted as described in
+[RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.txt).
+
+## Terminology
+
+`Sequence`: a 13-byte value formed by combining the current `Incarnation` of the database and the `Versionstamp` of the transaction. Sequences are monotonically increasing even when a database is relocated across FoundationDB clusters. See (RFC002)[LINK TBD] for a full explanation.
+
+`View Signature`: A md5 hash of the views, options, view language defined in a design document.
+
+---
+
+## Detailed Description
+
+CouchDB views are used to create secondary indexes in a database. An index is defined by creating map/reduce functions in a design document. This document describes building the map indexes on top of FoundationDB (FDB).
+
+An example map function is shown below:
+
+```json
+{
+  "_id": "_design/design-doc-id",
+  "_rev": "1-8d361a23b4cb8e213f0868ea3d2742c2",
+  "views": {
+    "map-view": {
+      "map": "function (doc) {\n  emit(doc._id, 1);\n}"
+    }
+  },
+  "language": "javascript"
+}
+```
+
+### Data model
+
+The data model for a map indexed is:
+
+```
+(<database>, ?DB_VIEWS, <view_signature>, ?VIEW_UPDATE_SEQ) = Sequence
+{<database>, ?DB_VIEWS, <view_signature>, ?VIEW_ID_INFO, view_id, ?VIEW_ROW_COUNT} = <row_count>
+{<database>, ?DB_VIEWS, <view_signature>, ?VIEW_ID_INFO, view_id, ?VIEW_KV_SIZE} = <kv_size>
+(<database>, ?DB_VIEWS, <view_signature>, ?VIEW_ID_RANGE, <_id>, <view_id>) = [total_keys, total_size, unique_keys]
+(<database>, ?DB_VIEWS, <view_signature>, ?VIEW_MAP_RANGE, <view_id>, <key>, <_id>, ?ROW_KEYS <count>) = <emitted_keys>
+(<database>, ?DB_VIEWS, <view_signature>, ?VIEW_MAP_RANGE, <view_id>, {<key>, <_id>}, <count>, ?ROW_VALUE) = <emitted_value>
+```
+
+Each field is defined as:
+
+- `<database>` is the specific database namespace
+- `?DB_VIEWS` is the views namespace.
+- `view_signature` is the design documents `View Signature`
```

Comment:
It'd be worth specifying whether this will be the same value as calculated for view sig by CouchDB 2.x. 

It's possible to imagine that, if it did, a migration path could stream direct from CouchDB 2.x index files into FoundationDB, so I think it's important to specify one way or the other.

Reply:
While its currently a direct copy of the logic I'm not sure its a good idea to assert the equivalence. I'm reasonably certain that the implementation doesn't touch on anything that's changed, but I could easily see a subtle change that does alter the calculation.